### DISCUSSION TITLE
Compile issue(s)

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#786] [Q&A]
Starting off on the wrong foot it seems:
```MODULE Main;
IMPORT IO;
BEGIN IO.Put("Hello World\n");
END Main.
```
```
import("libm3")
implementation("Main")
program("myprog")
```

```
~/programming/modula3/code/ cm3                                                                                                                                         
--- building in AMD64_LINUX ---

"/home/dnormandin/programming/modula3/code/m3makefile", line 3: quake runtime error: unable to open "/home/dnormandin/bin/cm3/pkg/io/AMD64_LINUX/.M3EXPORTS" for reading

--procedure--  -line-  -file---
import             --  <builtin>
include_dir         3  /home/dnormandin/programming/modula3/code/m3makefile
                    4  /home/dnormandin/programming/modula3/code/AMD64_LINUX/m3make.args

Fatal Error: package build failed
```
I'm on a 64bit Linux box. It compiled ok by simply doing `cm3' in the source directory, but I wanted to test using m3makefile. BTW, `cm3 -h' choked, telling me to use `-?'. Then it choked on that as well.

**Comment by ghost** [comment]
The only way I can replicate this error

```
"/home/dnormandin/programming/modula3/code/m3makefile", line 3: quake runtime error: unable to open "/home/dnormandin/bin/cm3/pkg/io/AMD64_LINUX/.M3EXPORTS" for reading
```

is with an m3makefile that looks like this:

```
import("io")
implementation("Main")
program("myprog")
```

That's the only thing I can find that makes any sense, because there is no `pkg/io` and there's no reason it would look for one unless given this particular invalid m3makefile as input.

So make sure you're saving your files before compiling, and that your inputs are what you think they are.

Also in the distribution you downloaded, there is an `examples/hello` that is basically this, so you could try building that instead.  If that works, then you can look for differences between that and your previous attempt.

(I'm going to hope that the lack of `MODULE Main;` was just a copy/paste error, because that would have resulted in quite a different error message.)

The failure with `cm3 -?` suggests that you're using a shell that treats `?` as a special character, perhaps `zsh`.  Try `cm3 -\?`.

**Reply by dukester** [reply]
OK - cm3 -\? solved the issue because I am using zsh. Thanks for that ...
I'm going to start over with the "hello  world" snippet, but I need clarification please.
The filename containing "Main" that I save - should that filename be identical to the MODULE name? As well, for a "non-Main" source file, should the filename be identical to the MODULE name? And in it I would EXPORT that module/filename?

The naming things is what's confusing me at the moment. Is it the same as Pascal and Oberon2? If it is, I can also look up their documentation.
Thx for your help.

**Reply by ghost** [reply]
Pascal can mean too many things.  But there are two big differences from Oberon.  First, interfaces are explicit and defined separately.  Second, there is not a one-to-one mapping between interfaces and implementations.  And yes, the filenames must match the module names.

So a simple module has an explicit interface in a separate file.

```modula-3
(* Hello.i3 *)
INTERFACE Hello;
END Hello.
```

```modula-3
(* Hello.m3 *)
MODULE Hello;
BEGIN
END Hello.
```

`Main` is special, it's just a marker interface that indicates that a module is a program.  To generate an executable, you have it implement `Main` somewhere.

```modula-3
(* Main.m3 *)
MODULE Main;
BEGIN
END Main.
```

But because there's no 1-to-1 mapping, a module can implement other interfaces.  It does this with an explicit `EXPORTS`.

```modula-3
(* Hello.m3 *)
MODULE Hello EXPORTS Main;
BEGIN
END Hello.
```

**Reply by dukester** [reply]
Excellent! Thanks for that and Merry Christmas season!

**Reply by dukester** [reply]
Got everything working OK! Thanks.
BTW, I too live "in the sticks" - in the foothills of the Canadian Rockies, SW of Calgary, Alberta. Wouldn't live in a concrete jungle ever again. Clean air, but a whack of grizzlies are always a challenge. Thanks again!

**Reply by ghost** [reply]
Sticks?  I'm in a thriving town of almost 300 people, and I haven't seen a bear all week.  :). Merry Christmas!

**Comment by RodneyBates** [comment]
Rats, I can't one-up you on this one, or perhaps one-down. My post office town is
386.  But I am 5 miles outside of it, in a county of 2500.

**Reply by dukester** [reply]
Complete with fiber-optic Internet? LOL
We're 1300 souls here in Turner Valley, but only a 20 min. drive from "no trees grow atop these mountains" Rockies. We _just_ got fiber-o not that long ago. Mountain lions occasionally drift into town from the Sheep River. Good place to learn Modula-3. :)

**Comment by dukester** [comment]
Follow-up question: 
cm3 -clean nukes everything - including the executable.
Is there a way to tell the compiler to save the executable somewhere other than the default location? TIA ...

**Comment by RodneyBates** [comment]
In most cases, cm3 -ship, after the compile and before the -clean,
puts an executable into /usr/local/cm3/bin.

But if you have compiled a compiler, this won't happen, rightly
so, because a lot of other stuff is needed, and also, you really
don't want to overlay your existing compiler, lest you get into a
state lacking either a chicken or an egg.

It takes something in scripts, but the way I am doing it,
you don't want to, without some other stuff.  Others will know
what to recommend.  But this is irrelevant unless you are working
on the compiler itself.

**Comment by dukester** [comment]
On Wed, 29 Dec 2021 09:47:28 -0800
"Rodney M. Bates" ***@***.***> wrote:

> In most cases, cm3 -ship, after the compile and before the -clean,
> puts an executable into /usr/local/cm3/bin.

Thank you sir! But just to be absolutely clear, I need to do:

cm3 -c <filename>
cm3 -ship
cm3 -clean

Is that the correct workflow?
--
Duke
** Text only please. Bottom post is best for me **
